File: packages/guards/src/ip-guard.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface IpGuardOptions {
  action: 'block' | 'warn' | 'mask';
  detectPrivate?: boolean;
  allowList?: string[];
  denyList?: string[];
}

const IPV4_RE = /\b(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})\b/g;
const IPV6_RE = /\b([0-9a-fA-F]{1,4}(:[0-9a-fA-F]{1,4}){7})\b/g;
const IPV6_SHORT_RE = /\b([0-9a-fA-F]{1,4}(:[0-9a-fA-F]{0,4}){2,7})\b/g;

interface IpMatch {
  ip: string;
  type: 'ipv4' | 'ipv6';
  isPrivate: boolean;
  start: number;
  end: number;
}

function isPrivateIpv4(ip: string): boolean {
  const parts = ip.split('.').map(Number);
  if (parts.some((p) => p < 0 || p > 255)) return false;
  if (parts[0] === 10) return true;
  if (parts[0] === 172 && parts[1] >= 16 && parts[1] <= 31) return true;
  if (parts[0] === 192 && parts[1] === 168) return true;
  if (parts[0] === 127) return true;
  if (parts[0] === 0) return true;
  return false;
}

function isValidIpv4(ip: string): boolean {
  const parts = ip.split('.').map(Number);
  return parts.length === 4 && parts.every((p) => p >= 0 && p <= 255);
}

function detect(text: string): IpMatch[] {
  const matches: IpMatch[] = [];
  const re4 = new RegExp(IPV4_RE.source, 'g');
  let m: RegExpExecArray | null;
  while ((m = re4.exec(text)) !== null) {
    if (!isValidIpv4(m[1])) continue;
    matches.push({
      ip: m[1],
      type: 'ipv4',
      isPrivate: isPrivateIpv4(m[1]),
      start: m.index,
      end: m.index + m[0].length,
    });
  }

  const re6 = new RegExp(IPV6_RE.source, 'g');
  while ((m = re6.exec(text)) !== null) {
    matches.push({
      ip: m[1],
      type: 'ipv6',
      isPrivate: m[1].startsWith('fe80') || m[1].startsWith('fc') || m[1].startsWith('fd'),
      start: m.index,
      end: m.index + m[0].length,
    });
  }

  return matches.sort((a, b) => b.start - a.start);
}

function maskIps(text: string, matches: IpMatch[]): string {
  let result = text;
  for (const m of matches) {
    result =
      result.slice(0, m.start) + '[IP_ADDRESS]' + result.slice(m.end);
  }
  return result;
}

export function ipGuard(options: IpGuardOptions): Guard {
  const detectPrivate = options.detectPrivate ?? true;

  return {
    name: 'ip-guard',
    version: '0.1.0',
    description: 'IP address detection, validation, and filtering',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(
      text: string,
      _ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      let matches = detect(text);

      if (!detectPrivate) {
        matches = matches.filter((m) => !m.isPrivate);
      }

      if (options.denyList) {
        matches = matches.filter((m) => options.denyList!.includes(m.ip));
      }

      if (options.allowList) {
        matches = matches.filter((m) => !options.allowList!.includes(m.ip));
      }

      const triggered = matches.length > 0;

      if (!triggered) {
        return {
          guardName: 'ip-guard',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      if (options.action === 'mask') {
        return {
          guardName: 'ip-guard',
          passed: true,
          action: 'override',
          overrideText: maskIps(text, matches),
          latencyMs: Math.round(performance.now() - start),
          details: {
            detected: matches.map(({ ip, type, isPrivate }) => ({
              ip, type, isPrivate,
            })),
          },
        };
      }

      return {
        guardName: 'ip-guard',
        passed: false,
        action: options.action,
        latencyMs: Math.round(performance.now() - start),
        details: {
          detected: matches.map(({ ip, type, isPrivate }) => ({
            ip, type, isPrivate,
          })),
        },
      };
    },
  };
}

File: python/open_guardrail/guards/ip_guard.py
"""IP address detection and blocking guard."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "ipv4": re.compile(
        r"\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}"
    ),
    "ipv6": re.compile(
        r"(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}"
        r"|(?:[0-9a-fA-F]{1,4}:){1,7}:"
        r"|::(?:[0-9a-fA-F]{1,4}:){0,6}[0-9a-fA-F]{1,4}"
    ),
}

_MASK_LABELS: dict[str, str] = {
    "ipv4": "[IP-ADDRESS]",
    "ipv6": "[IP-ADDRESS]",
}

_PRIVATE_RANGES = [
    re.compile(r"^10\."),
    re.compile(r"^172\.(1[6-9]|2\d|3[01])\."),
    re.compile(r"^192\.168\."),
    re.compile(r"^127\."),
]


def _is_private_ip(ip: str) -> bool:
    for pat in _PRIVATE_RANGES:
        if pat.match(ip):
            return True
    return False


def _detect(
    text: str, entities: list[str]
) -> list[dict]:
    matches: list[dict] = []
    for entity in entities:
        pat = _PATTERNS.get(entity)
        if not pat:
            continue
        for m in pat.finditer(text):
            matches.append(
                {
                    "type": entity,
                    "value": m.group(),
                    "start": m.start(),
                    "end": m.end(),
                }
            )
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask_text(text: str, matches: list[dict]) -> str:
    result = text
    for m in matches:
        label = _MASK_LABELS.get(m["type"], "[PII]")
        result = (
            result[: m["start"]] + label + result[m["end"] :]
        )
    return result


class _IpGuard:
    def __init__(
        self,
        *,
        action: str = "block",
        entities: Optional[List[str]] = None,
        private_only: bool = True,
    ) -> None:
        self.name = "ip-guard"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())
        self._private_only = private_only

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)

        if self._private_only:
            matches = [
                m for m in matches
                if m["type"] != "ipv4"
                or _is_private_ip(m["value"])
            ]

        elapsed = (time.perf_counter() - start) * 1000

        if not matches:
            return GuardResult(
                guard_name="ip-guard",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )

        detected = [
            {"type": m["type"], "value": m["value"]}
            for m in matches
        ]

        if self.action == "mask":
            return GuardResult(
                guard_name="ip-guard",
                passed=True,
                action="override",
                override_text=_mask_text(text, matches),
                latency_ms=round(elapsed, 2),
                details={"detected": detected},
            )

        return GuardResult(
            guard_name="ip-guard",
            passed=False,
            action=self.action,
            message="IP address detected in text",
            latency_ms=round(elapsed, 2),
            details={
                "detected": detected,
                "reason": (
                    "Text contains IP addresses"
                ),
            },
        )


def ip_guard(
    *,
    action: str = "block",
    entities: Optional[List[str]] = None,
    private_only: bool = True,
) -> _IpGuard:
    return _IpGuard(
        action=action,
        entities=entities,
        private_only=private_only,
    )
